src/coreclr/jit/llvm.cpp

```diff
@@ -1223,6 +1237,83 @@ void Llvm::fillPhis()
     }
 }
 
+void Llvm::buildUnaryOperation(genTreeOps oper, GenTree* node, Value* op1)
+{
+    Value* result;
+    switch (oper)
+    {
+        case GT_NEG:
+            if (op1->getType()->isFloatingPointTy())
+            {
+                result = _builder.CreateFNeg(op1, "fneg");
+            }
+            else
+            {
+                result = _builder.CreateNeg(op1, "neg");
+            }
+            break;
+        case GT_NOT:
+            result = _builder.CreateNot(op1, "not");
+            break;
+        default:
+            failFunctionCompilation();  // TODO-LLVM: other shift types
+    }
+    mapGenTreeToValue(node, result);
+}
+
+void Llvm::buildBinaryOperation(genTreeOps oper, GenTree* node, Value* op1, Value* op2)
+{
+    Value* result;
+    // widen short ints as required
+    op1 = zextIntIfNecessary(op1);
+    op2 = zextIntIfNecessary(op2);
```

Comment:
> genTreeOps oper, GenTree* node, Value* op1, Value* op2

I think we should reconsider this pattern, it loses information about small types (also `oper` is not necessary, since the node carries it).

There are basically two possibilities here:
1) Produce "normal" (upcast to `TYP_INT`, converted from pointers to integers) values when building the node.
2) Consume "normal" values when building the user.

The first is simpler, but will result in a more voluminous LLVM IR in case the user immediately downcasts, which may be confusing to read in the LLVM IR dumps. E. g. for this:
`CALL(short V00)` (where `short` is the signature type) it will produce:
```llvm
%argNorm = sext i16 %V00 to i32
%argVal  = trunc i32 %argNorm to i16
call %argVal
```
The code for the first approach would basically see us replace all `mapGenTreeToValue` with something like `produceValue`, which does the normalization.

The second approach would be to "normalize on demand", where the user nodes would pass in the types they "desire" (e. g. in the example above it'd be `short` for that call argument), this is basically what is going on here, only we would also pass the source node and the target type so as to not lose the information about small types:
```c++
Value* consumeValue(GenTree* node, var_types targetType);
Value* consumeValue(GenTree* node) => consumeValue(node, genActualType(node));
```
The primary benefit is it'll create less LLVM IR for calls and less confusing dumps, the primary downside is it is more complex (i. e. we'll have to to handle cases like `node->TypeGet()` is `TYP_BYTE`, `targetType` is `TYP_USHORT`, what does that mean - the answer being that it means we need one `sext i8 to i16` followed by `zext i16 to i32`).

Reply:
> `node->TypeGet()` is `TYP_BYTE`, `targetType` is `TYP_USHORT`, what does that mean - the answer being that it means we need one `sext i8 to i16`

Why is that `sext`?  `0xff TYP_BYTE || 0x0 TYP_BYTE`  with `targetType` is `TYP_USHORT`would then result in `0x0000ffff` wouldn't it ?  Is that right?

Reply:
> Why is that `sext`?

`TYP_BYTE` is signed, so it is `sext`. I guess my example is not the best because in the end we can just do `sext i8 to i16`, seeing as we're targeting `i16`, not `i32` - I got too caught up in the IR semantics for small types.

Reply:
Ok, thanks.  Let me redo it for the second approach.  Keeping the LLVM small and efficient is a goal for me as Mono already have a LLVM story, so the point here is to see if going this RyuJIT approach can match it or do better.  Anything that improves the chances of the LLVM being efficient is good for me.

Reply:
Well, I'm not sure now.  Having run the compiler again, I realize that a common cause of having to extend is the result of a comparison, e.g. `GT_EQ`.  In IR thats an `INT` type, but in LLVM its an `i1`.  In LLVM this is valid:

```
 %beq = icmp eq i32 %Loadloc0_, 23
  br i1 %beq, label %Block14, label %BlockD
```
(and is emitted by the IL backend)
In LLVM `i1` is the only type permitted by a conditional branch https://llvm.org/docs/LangRef.html#br-instruction so I haven't normalised the results of comparison operators in case they would just have to be `trunc`ed for the next instruction.  I can still do the suggestion, more or less, but it would be based on I suppose IR type, and then if that's the same, then the LLVM type of the operators.

Reply:
Would changing `targetType` to be an LLVM type solve the issue?

Reply:
>     if (node->TypeGet() != targetType)
>     {
>        // this is never hit
>        nodeValue = nodeValue;
>     }

Hmm, I don't think this is true. `ADD(IND short, IND byte)` would be the canonical example.

Of course now it is not hit because loads use the "normalize on produce" semantic:
https://github.com/dotnet/runtimelab/blob/be8db1725c4c3c5a35edd93f718193ace43c7b73/src/coreclr/jit/llvm.cpp#L1116-L1122

If we go with the "normalize on load" semantic (aka option 2), the `buildInd` code will be changed to produce `int16`s and `int8`s in terms of LLVM types. I do not think it would be a good idea to have a mix of normalize on load and normalize on produce, it seems be hard to reason about.

> Would it be better just to special case the LLVM/IR difference in logical op return values, ie i1 and fail on anything else?

Do we need to? In `consumeValue` we have, essentially, 3 types: the LLVM type of the value the node produces, the IR type and the target LLVM type.

Essentially, I think, the logic should be:
```c++
if (valueType != targetType) // Both are LLVM types
{
    assert(/* both are integral types less than or 32 bits in size */);
    if (sizeof(valueType) < sizeof(targetType))
    {
       // Upcast.
       if (varTypeIsSmall(node))
       {
           finalValue = varTypeIsSigned(node) ? sext(valueType to targetType) : zext(valueType to targetType);
       }
       else
       {
           // This is the special case for relops. Ordinary codegen "just knows" they need zero-extension.
           assert(valueType == i1);
           finalValue = zext(i1 to i32);
       }
    }
    else
    {
        // Truncate.
        finalValue = trunc(valueType to targetType);
    }
}
```
This should then handle all scenarios we care about: `JTRUE(RELOP)`, `CALL(small arg SMALL_NODE`, etc.

Reply:
Thanks for the code, I allowed int64 types so that I could use it for shifts also.

Reply:
Not sure I'd put the int32 -> in64 upcast there, seems prone to hiding type mismatches, but if you think that's better I have no objections.

Reply:
Agree, best not to conflate the issues.